python/cuml/cuml/experimental/accel/module_accelerator.py

```diff
@@ -0,0 +1,664 @@
+                loader=self,
+            ),
+        )
+
+        # ...but, we want to pretend like we expose the same attributes
+        # as the equivalent slow module
+        setattr(mod, "__dir__", slow_mod.__dir__)
+
+        # We set __path__ to the real path so that importers like
+        # jinja2.PackageLoader("slow_mod") work correctly.
+        # Note (dgd): this doesn't work for resources.files(data_module)
+        if getattr(slow_mod, "__path__", False):
+            assert mod.__spec__
+            mod.__path__ = slow_mod.__path__
+            mod.__spec__.submodule_search_locations = [*slow_mod.__path__]
+        return self._postprocess_module(mod, slow_mod, fast_mod)
+
+    @contextlib.contextmanager
+    def disabled(self):
+        """Return a context manager for disabling the module accelerator.
+
+        Within the block, any wrapped objects will instead deliver
+        attributes from their real counterparts (as if the current
+        nested block were in the denylist).
+
+        Returns
+        -------
+        Context manager for disabling things
+        """
+        try:
+            self._use_fast_lib_lock.acquire()
+            # The same thread might enter this context manager
+            # multiple times, so we need to remember the previous
+            # value
+            saved = self._use_fast_lib
+            self._use_fast_lib = False
+            yield
+        finally:
+            self._use_fast_lib = saved
+            self._use_fast_lib_lock.release()
+
+    @staticmethod
+    def getattr_real_or_wrapped(
+        name: str,
+        *,
+        real: dict[str, Any],
+        wrapped_objs,
+        loader: ModuleAccelerator,
+    ) -> Any:
+        """
+        Obtain an attribute from a module from either the real or
+        wrapped namespace.
+
+        Parameters
+        ----------
+        name
+            Attribute to return
+        real
+            Unwrapped "original" attributes
+        wrapped
+            Wrapped attributes
+        loader
+            Loader object that manages denylist and other skipping
+
+        Returns
+        -------
+        The requested attribute (either real or wrapped)
+        """
+        with loader._use_fast_lib_lock:
+            # Have to hold the lock to read this variable since
+            # another thread might modify it.
+            # Modification has to happen with the lock held for the
+            # duration, so if someone else has modified things, then
+            # we block trying to acquire the lock (hence it is safe to
+            # release the lock after reading this value)
+            use_real = not loader._use_fast_lib
+        if not use_real:
+            # Only need to check the denylist if we're not turned off.
+            frame = sys._getframe()
+            # We cannot possibly be at the top level.
+            assert frame.f_back
+            calling_module = pathlib.PurePath(frame.f_back.f_code.co_filename)
```

Comment:
Have we profiled to understand the performance implications of this mechanism? Inspecting the frame seems like something we should do only if we _really_ have to. In the context of cuML, we're already tracking whether or not we're internal to the cuML API, so do we need to use this?

Reply:
tl;dr: the time spent doing this is negligible. 

Just did a profiling of our importing times which was very illuminating after a long time of not doing it, and here is the current status quo of which I was a bit surprised:

- Time to import regular cuML: 11.21s
- Time to import cuml.experimental.accel (includes importing cuML): 12.389206647872925

Now, if we look further, here are the main components that take time to import: 

- UMAP 6.79s! of which: 
    - pynndescent 3.53
- cuML (without UMAP) 4.72s of which: 
    - cuDF 1.9s
    - CUDA contest initialization ~ 1s (estimated)
    - sklearn 0.9s
    - Dask 0.8s
- Rest 0.45s of which
    - 0.44 importing IPython
    - time spent inspecting frame (i.e. the actual question here): 0.003s


Gotta say, I was a bit surprised that the UMAP package is responsible for more than half our import time!